Receive updatable storage access prompt quirks from WebPrivacy

sysrqb opened the pull request:
#### f87f3abaf11709d4f12d8af4767f0fc6d728da6c
<pre>
Receive updatable storage access prompt quirks from WebPrivacy
<a href="https://bugs.webkit.org/show_bug.cgi?id=265798">https://bugs.webkit.org/show_bug.cgi?id=265798</a>
<a href="https://rdar.apple.com/119138065">rdar://119138065</a>

Reviewed by Chris Dumez and Wenson Hsieh.

This patch introduces a new structure that describes how to quirk multiple
sites within a single organization/company/entity by defining the top site and
subframe sites for which the storage access prompt should be shown. The set of
quirks is updatable via WebPrivacy.

This patch doesn&apos;t change any behavior, and it only plumbs the entries into the
NetworkProcess. I&apos;ll finish plumbing and making use of the data in a follow up
patch.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h: Added.
(WebCore::OrganizationStorageAccessPromptQuirk::isHashTableDeletedValue const):
(WebCore::OrganizationStorageAccessPromptQuirk::OrganizationStorageAccessPromptQuirk):
(WebCore::operator==):
(WebCore::add):
(WebCore::OrganizationStorageAccessPromptQuirkHashTraits::isEmptyValue):
(WebCore::OrganizationStorageAccessPromptQuirkHash::hash):
(WebCore::OrganizationStorageAccessPromptQuirkHash::equal):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::updateStorageAccessPromptQuirks):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::StorageAccessPromptQuirkObserver::create):
(WebKit::StorageAccessPromptQuirkObserver::invokeCallback):
(WebKit::StorageAccessPromptQuirkObserver::StorageAccessPromptQuirkObserver):
(WebKit::StorageAccessPromptQuirkController::cachedQuirks const):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForStorageAccessPromptQuirkChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::StorageAccessPromptQuirkController::shared):
(WebKit::StorageAccessPromptQuirkController::observeUpdates):
(WebKit::StorageAccessPromptQuirkController::setCachedQuirks):
(WebKit::StorageAccessPromptQuirkController::setCachedQuirksForTesting):
(WebKit::domainPairingsDictToMap):
(WebKit::StorageAccessPromptQuirkController::updateQuirks):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setStorageAccessPromptQuirkForTesting:withSubFrameDomains:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::NetworkProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setStorageAccessPromptQuirkForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/271748@main">https://commits.webkit.org/271748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61252ae59fbf90d480f96ea26c0b8b9a8be778bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29517 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32058 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26751 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5477 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26753 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25221 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5845 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29672 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33402 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26941 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26726 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7667 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26107 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36090 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7009 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6474 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7777 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h:
```diff
@@ -0,0 +1,78 @@
+        : organizationName { organizationName }, domainPairings { domainPairings } { }
+
+    OrganizationStorageAccessPromptQuirk(WTF::HashTableDeletedValueType)
+        : organizationName { WTF::HashTableDeletedValue } { }
```
Nit - we usually write it like this:

```suggestion
    OrganizationStorageAccessPromptQuirk(String organizationName, HashMap<RegistrableDomain, Vector<RegistrableDomain>> domainPairings)
        : organizationName { organizationName }
        , domainPairings { domainPairings }
        { }

    OrganizationStorageAccessPromptQuirk(WTF::HashTableDeletedValueType)
        : organizationName { WTF::HashTableDeletedValue }
        { }
```

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -94,6 +95,7 @@ @interface WKWebPrivacyNotificationListener : NSObject
 
 @implementation WKWebPrivacyNotificationListener {
     BlockPtr<void()> _linkFilteringDataCallback;
+    BlockPtr<void()> _StorageAccessPromptQuirksDataCallback;
```
```suggestion
    BlockPtr<void()> _storageAccessPromptQuirksDataCallback;
```

whsieh commented on Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h:
```diff
@@ -0,0 +1,78 @@
+
+    bool isHashTableDeletedValue() const { return organizationName.isHashTableDeletedValue(); }
+
+    OrganizationStorageAccessPromptQuirk(String organizationName, HashMap<RegistrableDomain, Vector<RegistrableDomain>> domainPairings)
```
Nit - this should either by a `HashMap<…>&&` or `const HashMap<…>&`.

whsieh commented on Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
```diff
@@ -67,6 +68,15 @@ typedef NS_ENUM(NSInteger, WPNetworkAddressVersion) {
+@interface WPStorageAccessPromptQuirksData : NSObject
+@property (nonatomic, readonly) NSArray<WPStorageAccessPromptQuirk *> *quirks;
+@end
+
```
Hm...I suspect this will break the internal build, against older versions of the WebPrivacy framework that are missing these symbols. You might need something like a new `HAVE(WEB_PRIVACY_STORAGE_ACCESS_PROMPT_QUIRK)` compile-time flag, or alternately, add a new `HAS_WEB_PRIVACY_STORAGE_ACCESS_PROMPT_QUIRK_CLASS` define in WP and add these forward declarations in the case where that's undefined.

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
 
+StorageAccessPromptQuirkController& StorageAccessPromptQuirkController::shared()
+{
+    static StorageAccessPromptQuirkController* sharedInstance = new StorageAccessPromptQuirkController;
```
`MainThreadNeverDestroyed` (from Chris' review comment in the [other PR](https://github.com/WebKit/WebKit/pull/21292/files)).

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+    RetainPtr options = adoptNS([PAL::allocWPResourceRequestOptionsInstance() init]);
+    [options setAfterUpdates:NO];
+
+    [[PAL::getWPResourcesClass() sharedInstance] requestStorageAccessPromptQuirksData:options.get() completionHandler:^(WPStorageAccessPromptQuirksData *data, NSError *error) {
```
This'll need a `-respondsToSelector:` check to avoid crashing, if the changes in WP haven't been submitted yet.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -642,6 +642,17 @@ - (void)_setResourceLoadStatisticsTestingCallback:(void (^)(WKWebsiteDataStore *
+
+    Vector<String> vectorSubSites;
+    for (NSString *subSite : subSites)
+        vectorSubSites.append({ subSite });
```
Nit - I think you can use `makeVector<String>(…)` here.

whsieh commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -1276,6 +1276,11 @@ void WebsiteDataStore::setResourceLoadStatisticsTimeAdvanceForTesting(Seconds ti
     protectedNetworkProcess()->setResourceLoadStatisticsTimeAdvanceForTesting(m_sessionID, time, WTFMove(completionHandler));
 }
 
+void WebsiteDataStore::setStorageAccessPromptQuirkForTesting(String topSite, Vector<String> subSites, CompletionHandler<void()>&& completionHandler)
```
Nit - not too critical since this is just testing code, but we usually pass around `Vector<String>&&` or `const Vector<String>&` to avoid unnecessary copying.

whsieh commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
```diff
@@ -164,6 +166,7 @@ class NetworkProcessProxy final : public AuxiliaryProcessProxy {
     void setGrandfathered(PAL::SessionID, const RegistrableDomain&, bool isGrandfathered, CompletionHandler<void()>&&);
     void setNotifyPagesWhenDataRecordsWereScanned(PAL::SessionID, bool, CompletionHandler<void()>&&);
     void setResourceLoadStatisticsTimeAdvanceForTesting(PAL::SessionID, Seconds, CompletionHandler<void()>&&);
+    void setStorageAccessPromptQuirkForTesting(String& topSite, Vector<String>& subSites, CompletionHandler<void()>&&);
```
Here, too — `const Vector<String>& subSites` or `Vector<String>&& subSites`.

whsieh commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -893,6 +903,16 @@ void NetworkProcessProxy::setResourceLoadStatisticsTimeAdvanceForTesting(PAL::Se
     sendWithAsyncReply(Messages::NetworkProcess::SetResourceLoadStatisticsTimeAdvanceForTesting(sessionID, time), WTFMove(completionHandler));
 }
 
+void NetworkProcessProxy::setStorageAccessPromptQuirkForTesting(String& topSite, Vector<String>& subSites, CompletionHandler<void()>&& completionHandler)
```
Ditto, re: the non-const ref.

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+void StorageAccessPromptQuirkController::updateQuirks(CompletionHandler<void()>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    if (!WebKit::canUseWebPrivacyFramework()) {
```
```suggestion
    if (!WebKit::canUseWebPrivacyFramework() || ![PAL::getWPResourcesClass() instancesRespondToSelector:@selector(requestStorageAccessPromptQuirksData:completionHandler:)]) {
```

cdumez commented on Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h:
```diff
@@ -0,0 +1,81 @@
+
+struct OrganizationStorageAccessPromptQuirkHashTraits : SimpleClassHashTraits<OrganizationStorageAccessPromptQuirk> {
+    static const bool hasIsEmptyValueFunction = true;
+    static const bool emptyValueIsZero = false;
```
I *think* we can use `true` here no? It would be more efficient and AFAIK, memset(0) would do the right thing for String and HashMap.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,46 @@ class LinkDecorationFilteringController {
+    {
+    }
+
+    static Ref<StorageAccessPromptQuirkObserver> create(Function<void()>&& callback)
```
I think this class would look better with this public and the invokeCallback() too, as mentioned in my earlier review. Without a "friend".

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+        m_notificationListener = adoptNS([WKWebPrivacyNotificationListener new]);
+        [m_notificationListener listenForStorageAccessPromptQuirkChanges:^{
+            updateQuirks([this] {
+                for (auto& observer : m_observers)
```
I think you want to use `WeakHashSet::forEach()` here to make a copy of the observers. It is a bit risky otherwise as an observer could unregister / register itself during the execution of the callback.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+        Vector<WebCore::RegistrableDomain> subFrameDomains;
+        for (NSString *subFrameDomain : [domainPairings objectForKey:topDomain])
+            subFrameDomains.append(WebCore::RegistrableDomain::fromRawString(subFrameDomain));
+        map.set(WebCore::RegistrableDomain::fromRawString(String { topDomain }), WTFMove(subFrameDomains));
```
Why is this still a `set()` call instead of `add()`, as mentioned in my earlier review.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -239,6 +249,81 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
+        return;
+    }
+
+    static MainThreadNeverDestroyed<Vector<CompletionHandler<void()>, 1>> lookupCompletionHandler;
```
`lookupCompletionHandlers` plural since it's a vector.

cdumez commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -856,6 +856,11 @@ void NetworkProcess::setResourceLoadStatisticsTimeAdvanceForTesting(PAL::Session
     completionHandler();
 }
 
+void NetworkProcess::setStorageAccessPromptQuirkForTesting(String, Vector<String>, CompletionHandler<void()>&& completionHandler)
```
Vector and String should not be passed by value here.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1617,6 +1637,16 @@ void NetworkProcessProxy::increaseQuota(PAL::SessionID sessionID, const WebCore:
+    if (StorageAccessPromptQuirkController::shared().cachedQuirks().isEmpty())
+        return;
+
+    send(Messages::NetworkProcess::SetStorageAccessPromptQuirks(StorageAccessPromptQuirkController::shared().cachedQuirks()), 0);
```
We should probably just call this `UpdateStorageAccessPromptQuirks`, if it is only used for updating.

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -244,6 +246,14 @@ NetworkProcessProxy::NetworkProcessProxy()
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->sendStorageAccessPromptQuirksData();
+    });
+    sendStorageAccessPromptQuirksData();
```
`updateStorageAccessPromptQuirksData()` since it should only be used for updates?

cdumez commented on Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
```diff
@@ -1617,6 +1637,16 @@ void NetworkProcessProxy::increaseQuota(PAL::SessionID sessionID, const WebCore:
+void NetworkProcessProxy::sendStorageAccessPromptQuirksData()
+{
+#if ENABLE(ADVANCED_PRIVACY_PROTECTIONS)
+    if (StorageAccessPromptQuirkController::shared().cachedQuirks().isEmpty())
```
What if the quirks get updated so that we go from a non-empty list to an empty list? Isn't that a possibility (i.e. we no longer need the quirks). It seems we would fail to update the WebProcess's quirks in this case no?

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,46 @@ class LinkDecorationFilteringController {
 
+class StorageAccessPromptQuirkObserver : public RefCounted<StorageAccessPromptQuirkObserver>, public CanMakeWeakPtr<StorageAccessPromptQuirkObserver> {
+public:
+    ~StorageAccessPromptQuirkObserver() = default;
```
nit: would be nicer if this came *after* the create() function.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,46 @@ class LinkDecorationFilteringController {
+    void invokeCallback() { m_callback(); }
+
+private:
+
```
I don't think we usually have a blank line here.

cdumez commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
```diff
@@ -87,6 +88,46 @@ class LinkDecorationFilteringController {
+class StorageAccessPromptQuirkController {
+public:
+    static StorageAccessPromptQuirkController& shared();
+    StorageAccessPromptQuirkController() = default;
```
Should be private (marking NeverDestroyed as friend).

sysrqb committed:
Receive updatable storage access prompt quirks from WebPrivacy
https://bugs.webkit.org/show_bug.cgi?id=265798
rdar://119138065

Reviewed by Chris Dumez and Wenson Hsieh.

This patch introduces a new structure that describes how to quirk multiple
sites within a single organization/company/entity by defining the top site and
subframe sites for which the storage access prompt should be shown. The set of
quirks is updatable via WebPrivacy.

This patch doesn't change any behavior, and it only plumbs the entries into the
NetworkProcess. I'll finish plumbing and making use of the data in a follow up
patch.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.h:
* Source/WebCore/PAL/pal/cocoa/WebPrivacySoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/network/OrganizationStorageAccessPromptQuirk.h: Added.
(WebCore::OrganizationStorageAccessPromptQuirk::isHashTableDeletedValue const):
(WebCore::OrganizationStorageAccessPromptQuirk::OrganizationStorageAccessPromptQuirk):
(WebCore::operator==):
(WebCore::add):
(WebCore::OrganizationStorageAccessPromptQuirkHashTraits::isEmptyValue):
(WebCore::OrganizationStorageAccessPromptQuirkHash::hash):
(WebCore::OrganizationStorageAccessPromptQuirkHash::equal):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::updateStorageAccessPromptQuirks):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.h:
(WebKit::StorageAccessPromptQuirkObserver::create):
(WebKit::StorageAccessPromptQuirkObserver::invokeCallback):
(WebKit::StorageAccessPromptQuirkObserver::StorageAccessPromptQuirkObserver):
(WebKit::StorageAccessPromptQuirkController::cachedQuirks const):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(-[WKWebPrivacyNotificationListener listenForStorageAccessPromptQuirkChanges:]):
(-[WKWebPrivacyNotificationListener didUpdate:]):
(WebKit::StorageAccessPromptQuirkController::shared):
(WebKit::StorageAccessPromptQuirkController::observeUpdates):
(WebKit::StorageAccessPromptQuirkController::setCachedQuirks):
(WebKit::StorageAccessPromptQuirkController::setCachedQuirksForTesting):
(WebKit::domainPairingsDictToMap):
(WebKit::StorageAccessPromptQuirkController::updateQuirks):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _setStorageAccessPromptQuirkForTesting:withSubFrameDomains:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeNetworkProcess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::NetworkProcessProxy):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::setStorageAccessPromptQuirkForTesting):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/271748@main
